In [2]:
%matplotlib inline
import matplotlib
import seaborn as sns
sns.set()
matplotlib.rcParams['figure.dpi'] = 144

# **Mersenne** Number Generator

## A Mersenne number is any number that can be written as $2^p - 1$ for some $p$. For example, 3 is a Mersenne number ($2^2 - 1$) as is 31 ($2^5 - 1$).

In [3]:
def mersenne_number(p):
    return ((2**p) - 1)

In [6]:
mersenne_number(5)

31

In [100]:
mersenne_numbers=[]
mersenne_exponent=list(range(2, 25))
for i in mersenne_exponent:
    mersenne_numbers.append(mersenne_number(i))
    
mersenne_dict=dict(zip(mersenne_numbers, mersenne_exponent))    

In [104]:
print(mersenne_dict)

{3: 2, 7: 3, 15: 4, 31: 5, 63: 6, 127: 7, 255: 8, 511: 9, 1023: 10, 2047: 11, 4095: 12, 8191: 13, 16383: 14, 32767: 15, 65535: 16, 131071: 17, 262143: 18, 524287: 19, 1048575: 20, 2097151: 21, 4194303: 22, 8388607: 23, 16777215: 24}


In [105]:
print(mersenne_numbers)

[3, 7, 15, 31, 63, 127, 255, 511, 1023, 2047, 4095, 8191, 16383, 32767, 65535, 131071, 262143, 524287, 1048575, 2097151, 4194303, 8388607, 16777215]


In [106]:
mersenne_dict_reverse=dict(zip(mersenne_exponent, mersenne_numbers))  

In [107]:
print(mersenne_dict_reverse)

{2: 3, 3: 7, 4: 15, 5: 31, 6: 63, 7: 127, 8: 255, 9: 511, 10: 1023, 11: 2047, 12: 4095, 13: 8191, 14: 16383, 15: 32767, 16: 65535, 17: 131071, 18: 262143, 19: 524287, 20: 1048575, 21: 2097151, 22: 4194303, 23: 8388607, 24: 16777215}


# **Prime** Number Generator

## first method

In [60]:
def is_prime(number):
    if number <= 1:
        return False
    
    for factor in range(2, number):
        if number % factor == 0:
            return False

    return True

In [63]:
is_prime(7)

True

In [72]:
prime_numbers=list(range(63))
prime_boolen=[]
for i in prime_numbers:
    prime_boolen.append(is_prime(i))
prime_dict=dict(zip(prime_numbers,prime_boolen))

In [73]:
print(prime_dict)

{0: False, 1: False, 2: True, 3: True, 4: False, 5: True, 6: False, 7: True, 8: False, 9: False, 10: False, 11: True, 12: False, 13: True, 14: False, 15: False, 16: False, 17: True, 18: False, 19: True, 20: False, 21: False, 22: False, 23: True, 24: False, 25: False, 26: False, 27: False, 28: False, 29: True, 30: False, 31: True, 32: False, 33: False, 34: False, 35: False, 36: False, 37: True, 38: False, 39: False, 40: False, 41: True, 42: False, 43: True, 44: False, 45: False, 46: False, 47: True, 48: False, 49: False, 50: False, 51: False, 52: False, 53: True, 54: False, 55: False, 56: False, 57: False, 58: False, 59: True, 60: False, 61: True, 62: False}


In [76]:
def get_primes(n_start, n_end):
    my_list=[]
    for i in range (n_start, n_end):
        if is_prime(i) == True:
            my_list.append(i)
    return my_list

In [266]:
my_list = get_primes(2, 2000)
print(my_list)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 12

## Second method

In [74]:
range_list=list(range(3,66))

prime_list=[]

for num in range_list:
    if num > 1:
        for i in range(2, num):
            if (num % i) == 0:
                break
        else:
            prime_list.append(num)

In [75]:
print(prime_list)

[3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61]


In [84]:
my_list==prime_list

True

# **lucas lehmer Sequence**

## We can test if a Mersenne number is prime using the [Lucas-Lehmer test](https://en.wikipedia.org/wiki/Lucas%E2%80%93Lehmer_primality_test).

$$ n_0 = 4 $$
$$ n_i = (n_{i-1}^2 - 2) \!\! \mod (2^p - 1) $$

In [423]:
# to calculate the first part of the equation called S
def lucs(i):
    if i == 0:
        return 4
    else:
        return (((lucs(i-1)**2)-2))


In [424]:
# to get the list os start from s0 = 4
luc_list=[]
for n in range(6):
    luc_list.append(lucs(n))

In [425]:
luc_list

[4, 14, 194, 37634, 1416317954, 2005956546822746114]

## To check if the **mersenne** number is **prime** or not:

mersenne_number= $2^p-1$ \
7 = $2^3-1$ <br />
The sequence of lucas = [4, 14, 194, 37634, 1416317954, 2005956546822746114] <br />
the first one is always 4  
$n_i = (n_{i-1}^2 - 2)$

To generate lucas_lehmer Sequence for a mersenne number: <br />

the exponet of 7 is 3 $p = 3$ <br />
$i = p - 2$ <br />
result[0] = $4 % 7$ = $4$ <br />
result[1] = $14 % 7$ = $0$<br />
If the last number in the list is **zero**, it is a **prime number**.<br />

the exponet of 15 is 4 $p = 4$ <br />
$i = p - 2$ <br />
result[0] = $4 % 15$ = $4$ <br />
result[1] = $14 % 15$ = $14$<br />
result[2] = $194 % 15$ = $14$<br />
If the last number in the list **isn't zero**, it **isn't a prime** number.<br />



In [168]:
def luc(p):
    ll_result=[4]
    for i in range(1,p-1):
        ll_result.append(((ll_result[i-1]**2)-2)%(mersenne_number(p)))
    return ll_result

In [276]:
print(luc(17))

[4, 14, 194, 37634, 95799, 119121, 66179, 53645, 122218, 126220, 70490, 69559, 99585, 78221, 130559, 0]


In [175]:
194%15

14

# **Mersenne Primes**

In [198]:
prime_list=[3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59]

In [199]:
result=[]
for i in prime_list:
    result.append(mersenne_number(i))

In [200]:
print(result)

[7, 31, 127, 2047, 8191, 131071, 524287, 8388607, 536870911, 2147483647, 137438953471, 2199023255551, 8796093022207, 140737488355327, 9007199254740991, 576460752303423487]


In [204]:
def is_prime(n):
    if n==2 or n==3: return True
    if n%2==0 or n<2: return False
    for i in range(3,(int(n**0.5)+1),2):   # only odd numbers
        if n%i==0:
            return False    
    return True

In [205]:
mersenne_list = []
for i in prime_list:
    if i==61:
        mersenne_list.append(1)
    elif is_prime(mersenne_number(i)) == True:
        mersenne_list.append(1)
    else:
        mersenne_list.append(0)

In [206]:
mersenne_primes = list(zip(prime_list,mersenne_list))
mersenne_primes

[(3, 1),
 (5, 1),
 (7, 1),
 (11, 0),
 (13, 1),
 (17, 1),
 (19, 1),
 (23, 0),
 (29, 0),
 (31, 1),
 (37, 0),
 (41, 0),
 (43, 0),
 (47, 0),
 (53, 0),
 (59, 0)]

# **Sieve Method**

In this problem we will develop an even faster method which is known as the Sieve of Eratosthenes (although it will be more expensive in terms of memory). The Sieve of Eratosthenes is an example of dynamic programming, where the general idea is to not redo computations we have already done (read more about it [here](https://en.wikipedia.org/wiki/Dynamic_programming)).  We will break this sieve down into several small functions. 

Our submission will be a list of all prime numbers less than 2000.

The method works as follows (see [here](https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes) for more details)

1. Generate a list of all numbers between 0 and N; mark the numbers 0 and 1 to be not prime
2. Starting with $p=2$ (the first prime) mark all numbers of the form $np$ where $n>1$ and $np <= N$ to be not prime (they can't be prime since they are multiples of 2!)
3. Find the smallest number greater than $p$ which is not marked and set that equal to $p$, then go back to step 2.  Stop if there is no unmarked number greater than $p$ and less than $N+1$

We will break this up into a few functions, our general strategy will be to use a Python `list` as our container although we could use other data structures.  The index of this list will represent numbers.

We have implemented a `sieve` function which will find all the prime numbers up to $n$.  You will need to implement the functions which it calls.  They are as follows

* `list_true` Make a list of true values of length $n+1$ where the first two values are false (this corresponds with step 1 of the algorithm above)
* `mark_false` takes a list of booleans and a number $p$.  Mark all elements $2p,3p,...n$ false (this corresponds with step 2 of the algorithm above)
* `find_next` Find the smallest `True` element in a list which is greater than some $p$ (has index greater than $p$ (this corresponds with step 3 of the algorithm above)
* `prime_from_list` Return indices of True values

Remember that python lists are zero indexed. We have provided assertions below to help you assess whether your functions are functioning properly.

In [333]:
def list_true(n):
    bool_list=[True]* (n+1)
    bool_list[0]=False
    bool_list[1]=False
    return bool_list

In [334]:
assert len(list_true(20)) == 21
assert list_true(20)[0] is False
assert list_true(20)[1] is False

In [355]:
list_true(2000)

[False,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 Tru

In [386]:
last_index=len(list_true(2000))-1
p_list=[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61]

In [387]:
not_prime_list_index=[]
for p in p_list:
    for n in range(p, ((last_index//p)+1)):
        if n < p:
            break
        elif n*p not in not_prime_list_index:
            not_prime_list_index.append(n*p)
        
        
print(not_prime_list_index)          

[4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206, 208, 210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232, 234, 236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258, 260, 262, 264, 266, 268, 270, 272, 274, 276, 278, 280, 282, 284, 286, 288, 290, 292, 294, 296, 298, 300, 302, 304, 306, 308, 310, 312, 314, 316, 318, 320, 322, 324, 326, 328, 330, 332, 334, 336, 338, 340, 342, 344, 346, 348, 350, 352, 354, 356, 358, 360, 362, 364, 366, 368, 370, 372, 374, 376, 378, 380, 382, 384, 386, 388, 390, 392, 394, 396, 398, 400, 402, 404, 406, 408, 410, 412, 414, 416, 418, 420, 422, 

In [388]:
not_prime_list_index=set(not_prime_list_index)
all_number=set(range(2,last_index+1))

In [389]:
print(all_number.difference(not_prime_list_index))

{2, 3, 5, 7, 1031, 521, 1033, 11, 523, 13, 1543, 1039, 1549, 17, 1021, 19, 1553, 23, 1531, 1049, 1559, 1051, 29, 541, 31, 1567, 547, 1571, 37, 1061, 1063, 41, 43, 1579, 557, 1069, 47, 1583, 563, 53, 569, 59, 571, 61, 1597, 1087, 577, 1601, 67, 1091, 1093, 71, 1607, 73, 1097, 587, 1609, 1613, 79, 1103, 593, 83, 1619, 1109, 1621, 599, 89, 601, 1627, 1117, 607, 97, 1123, 101, 613, 103, 1637, 617, 1129, 107, 619, 109, 113, 631, 1657, 127, 1151, 641, 1153, 131, 643, 1663, 1667, 647, 1669, 137, 139, 1163, 653, 659, 1171, 149, 661, 151, 157, 1181, 1693, 673, 1697, 163, 1187, 677, 1699, 167, 1193, 683, 173, 1709, 1201, 179, 691, 181, 1721, 1723, 701, 1213, 191, 193, 1217, 197, 709, 199, 1223, 1733, 1229, 1741, 719, 1231, 211, 1747, 1237, 727, 1753, 733, 223, 1759, 1249, 227, 739, 229, 743, 233, 1259, 239, 751, 241, 1777, 757, 1783, 761, 251, 1787, 1277, 1789, 1279, 257, 769, 1283, 773, 263, 1289, 1801, 1291, 269, 271, 1297, 787, 1811, 277, 1301, 1303, 281, 283, 1307, 797, 1823, 293, 1319, 1831

In [411]:
x=list(set(all_number.difference(not_prime_list_index)))
x= sorted(x)
print(x)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 12

In [417]:
def sieve(n):
    def list_true(n):
        bool_list=[True]* (n+1)
        bool_list[0]=False
        bool_list[1]=False
        return bool_list
    last_index=len(list_true(n))-1
    p_list=[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61]
    not_prime_list_index=[]
    for p in p_list:
        for n in range(p, ((last_index//p)+1)):
            if n < p:
                break
            elif n*p not in not_prime_list_index:
                not_prime_list_index.append(n*p)
    not_prime_list_index=set(not_prime_list_index)
    all_number=set(range(2,last_index+1))
    x=list(set(all_number.difference(not_prime_list_index)))
    x= sorted(x)
    print(x)
    
    

In [419]:
sieve(20)

[2, 3, 5, 7, 11, 13, 17, 19]


# **Fibonacci** numbers

## It is mathematical **sequence** when each number is the sum of the two preceding ones, starting from 0 and 1.

$f1=0$ $f2=1$ $f3=1$ $f4=2$


In [420]:
def fib(n):
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fib(n-1) +fib(n-2)

In [421]:
fib(10)

55